pkg/vere/disk.c

```diff
@@ -1019,66 +1031,415 @@ u3_disk_init(c3_c* pax_c, u3_disk_cb cb_u)
+  if ( 0 != c3_rmdir(epo_c) ) {
+    fprintf(stderr, "disk: failed to delete epoch directory\r\n");
+    return c3n;
+  }
+
+  //  cleanup
+  u3_dire_free(dir_u);
+
+  //  success
+  return c3y;
+}
+
+/* u3_disk_epoc_last: get latest epoch number.
+*/
+c3_o
+u3_disk_epoc_last(u3_disk* log_u, c3_d* lat_d)
+{
+  c3_o ret_o = c3n;  //  return c3n if no epoch directories exist
+  *lat_d = 0;        //  initialize lat_d to 0
+  u3_dire* die_u = u3_foil_folder(log_u->com_u->pax_c);
+  u3_dent* den_u = die_u->dil_u;
+  while ( den_u ) {
+    c3_d epo_d = 0;
+    if ( 1 == sscanf(den_u->nam_c, "0i%" PRIc3_d, &epo_d) ) {
+      ret_o = c3y;   //  NB: returns yes if the directory merely exists
+      *lat_d = c3_max(epo_d, *lat_d);  //  update the latest epoch number
+    }
+    den_u = den_u->nex_u;
+  }
+
+  u3_dire_free(die_u);
+
+  return ret_o;
+}
+
+/* u3_disk_epoc_vere: get binary version from epoch.
+*/
+c3_o
+u3_disk_epoc_vere(u3_disk* log_u, c3_d epo_d, c3_c* ver_w)
+{
+  struct stat buf_u;
+  c3_c* ver_c;
+  c3_w red_w, len_w;
+  c3_i ret_i, fid_i;
+  ret_i = asprintf(&ver_c, "%s/0i%" PRIc3_d "/vere.txt",
+           log_u->com_u->pax_c, epo_d);
+  u3_assert( ret_i > 0 );
+
+  fid_i = c3_open(ver_c, O_RDONLY, 0644);
+
+  if ( (fid_i < 0) || (fstat(fid_i, &buf_u) < 0) ) {
+    fprintf(stderr, "disk: failed to open vere.txt in epoch 0i%" PRIc3_d
+                    "\r\n", epo_d);
+    return c3n;
+  }
+
+  len_w = buf_u.st_size;
+  red_w = read(fid_i, ver_w, len_w);
+  close(fid_i);
+
+  if ( len_w != red_w ) {
+    fprintf(stderr, "disk: failed to read vere.txt in epoch 0i%" PRIc3_d
+                    "\r\n", epo_d);
+    return c3n;
+  }
+
+  //  trim trailing whitespace
+  ver_w[len_w] = 0;
+  while ( len_w-- && isspace(ver_w[len_w]) ) {
+    ver_w[len_w] = 0;
+  }
+
+  return c3y;
+}
+
+/* u3_disk_migrate: migrates disk format.
+ */
+c3_o u3_disk_migrate(u3_disk* log_u)
+{
+  /*  migration steps (* indicates breakpoint set):
+   *  0. detect whether we need to migrate or not
+   *     a. if it's a fresh boot via u3_Host.ops_u.nuu -> skip migration (returns yes)
+   *     b. if data.mdb is readable in log directory -> execute migration (returns yes or no)
+   *        if not -> skip migration (returns yes)
+   *  1. initialize epoch 0i0 (first call to u3_disk_epoc_init()) *
+   *     a. creates epoch directory
+   *     b. creates epoch version file
+   *     c. creates binary version file
+   *     d. initializes database
+   *     e. reads metadata from old database
+   *     f. writes metadata to new database
+   *     g. loads new epoch directory and sets it in log_u
+   *  2. create hard links to data.mdb and lock.mdb in 0i0/
+   *  3. rollover to new epoch (second call to u3_disk_epoc_init()) *
+   *     a. same as 1a-g but also copies current snapshot between c/d steps
+   *  4. delete backup snapshot (c3_unlink() and c3_rmdir() calls) *
+   *  5. delete old data.mdb and lock.mdb files (c3_unlink() calls) *
```

Comment:
The "(* indicates breakpoint set" and "*" instances, you'll want to probably remove since they were for debugging purposes